First batch of snippet improvements. (#52)
xsleaks/wiki: content/docs/attacks/navigations.md

## Download Trigger

When an endpoint sets the [`Content-Disposition: attachment`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) header, it instructs the browser to download the response as an attachment instead of navigating to it. Detecting if this behavior occurred might allow attackers to leak private information if outcome depends on the state of the victim's account.

### Download bar

In Chromium-based browsers when a file is downloaded, a preview of the download process appears in a bar at the bottom, integrated into the browser window. By monitoring the window height attackers could detect whether the "download bar" opened.

```javascript
// Read the current height of the window
var screenHeight = window.innerHeight;

> // Load the page that will or will not trigger the download

Reviewer: ```suggestion
// Load the page that may or may not trigger the download
```

window.open('https://example.org');

> // Wait for a tab to load

Reviewer: ```suggestion
// Wait for the tab to load
```

setTimeout(() => {
    // If the download bar appears, the height of all tabs will be smaller

>     if (window.innerHeight !== screenHeight) {

Reviewer: nit: so the `if` matches the comment above it

```suggestion
    if (window.innerHeight < screenHeight) {
```

      console.log('Download bar detected');
    } else {
      console.log('Download bar not detected');
    }
}, 2000);
```

{{< hint warning >}}

> This attack is only possible in Chromium-based browsers with automatic downloads enabled. The attack can't be also repeated since the user needs to close the download bar for it to be measureable again.

Reviewer: ```suggestion
This attack is only possible in Chromium-based browsers with automatic downloads enabled. The attack can't be also repeated since the user needs to close the download bar for it to be measurable again.
```

{{< /hint >}}

### Download Navigation (with iframes)

Another way to test for the [`Content-Disposition: attachment`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) header is to check if a navigation occurred. If a page load causes a download, it will not trigger a navigation and the window will stame within the same origin.

> The following snippet can be used to detect whether such navigation has occured.

Reviewer: ```suggestion
The following snippet can be used to detect whether such a navigation has occurred.
```

```javascript
// Set the destination URL
var url = 'https://example.org';
// Create an outer iframe
var iframe = document.createElement('iframe');
document.body.appendChild(iframe);
iframe.srcdoc = `

>   <!-- Create an inner iframe -->

Reviewer: I think it would be useful to explain here why we're creating a nested iframe (it isn't immediately obvious to me)

Author: I managed to improve the snippet further in https://github.com/xsleaks/wiki/pull/52/commits/73d56852bb81efa3a4e3c0519071eb7cc774f77e 

Let me know if that seems better. Plus, removed the step-by-step explainer at cost of one more hint why we used inner and outer iframes.

Reviewer: LGTM :+1:

  <iframe name="inner" src="${url}" ></iframe>
  <script>
    // When window loads check if the inner iframe is same origin.
    onload = () => {
      try {

>           inner.name; // if the navigations occurs it will throw an exception

Reviewer: ```suggestion
          // If a navigation occurs, the iframe will be cross-origin so accessing "inner.name" will throw an exception
          inner.name;
```

          parent.console.log('Download attempt detected');
      } catch(e) {
          parent.console.log('No download attempt detected');
      }
    }
  </script>`;
```

{{< hint info >}}

> This attack will work regardless of the [Framing Protections]({{< ref "xfo" >}}), because the `X-Frame-Options` and `Content-Security-Policy` headers are ignored in case of `Content-Disposition` being present.

Reviewer: ```suggestion
This attack will work regardless of any [Framing Protections]({{< ref "xfo" >}}), because the `X-Frame-Options` and `Content-Security-Policy` headers are ignored if `Content-Disposition: attachment` is specified.
```

{{< /hint >}}

The technique works as follows:

> 1. Include an inner `iframe` inside an outer `iframe` and point it to the destination resource.

Reviewer: This bit seems like it may be redundant? What do you think? If the demo JS is clear enough, I think we could cut this out (just like how we don't have it for the bit about the download bar).

2. If the embedded resource triggers a download the origin of the inner `iframe` will remain as its parent, otherwise, it will change to `https://example.org`.

> 3. Because the `window.onload` event listener triggers when all the resources has been loaded (including iframes), no timing is needed to determine when the iframe finished loading.

Reviewer: Reorganizing this sentence so it flows similarly to other sentences

```suggestion
3. The outer iframe can detected when the inner iframe has finished loading using `window.onload` which triggers when all resources have finished loading
```

4. When the window loaded, the outer iframes tests whether the inner iframe stayed the same origin.

### Download Navigation (without iframes)

> A variation of the technique presented in the previous section can be also effectively tested using windows.

Reviewer: To clarify it for people like me who first jumped to Windows the OS :upside_down_face: 

```suggestion
A variation of the technique presented in the previous section can also be effectively tested using `window` objects.
```

```javascript
// Set the destination URL
var url = 'https://example.org';
// Get a window reference

> var win = windowp.open(url);

Reviewer: ```suggestion
var win = window.open(url);
```

// Wait for the window to load.
setTimeout(() => {
      try {

>           win.name; // if the navigations occurs it will throw an exception

Reviewer: ```suggestion
          // If a navigation occurs, the iframe will be cross-origin so accessing "win.name" will throw an exception
          win.name;
```

          parent.console.log('Download attempt detected');
      } catch(e) {
          parent.console.log('No download attempt detected');
      }
}, 2000);
```

## Server-Side Redirects

### Inflation

A server-side redirect can be detected from a cross-origin page if the destination URL increases in size and contains an attacker controlled input (either in the form of a query string parameter or a path). The following technique relies on the fact that it is possible to induce an error in most web-servers by generating big requests parameters/paths. Since the redirect increases the size of the URL, it can be detected by sending exactly one character less than the server maximum capacity. That way if the size increases the server will respond with an error that can be detected from a cross-origin page (eg via Error Events).

{{< hint info >}}

> The example of this attack was implemented [here](https://xsleaks.github.io/xsleaks/examples/redirect/).

Reviewer: ```suggestion
An example of this attack can be seen [here](https://xsleaks.github.io/xsleaks/examples/redirect/).
```

{{< /hint >}}
## Cross-Origin Redirects

### CSP Violations

[Content-Security-Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) (CSP) is an in-depth defense mechanism against XSS and data injection attacks. When a CSP is violated, a `SecurityPolicyViolationEvent` is thrown. An attacker can set up a CSP which will trigger a `Violation` event every time a `fetch` follows an URL not set in the CSP directive. This will allow an attacker to detect if a redirect to another origin occurred [^2] [^3].

The example below will trigger a `SecurityPolicyViolationEvent` if the website set in fetch API (line 6) redirects to a website different then `target.page`.

{{< highlight html "linenos=table,linenostart=1" >}}
<!-- Set the Content-Security-Policy to only allow example.org -->
<meta http-equiv="Content-Security-Policy"
      content="connect-src https://example.org">
<script>
// Listen for a CSP violation event
document.addEventListener('securitypolicyviolation', () => {

>   console.log("redirected");

Reviewer: ```suggestion
  console.log("Detected a redirect to somewhere other than example.org");
```
